### ISSUE TITLE
迁移训练收敛很慢问题

### ISSUE THREAD

**Issue by hei-my** [zzh8829/yolov3-tf2#339]
你好，我使用labelme标注工具标注一千张图片后，只预测一个类别。然后将图片放到JPEGImages中，标注放到Annotations中，然后将1000个文件拆分为test.txt，train.txt，trainval.txt，val.txt中，其中每个文件都是如下只包含图片的名称。
`0
1
10
100
1000
......
......
......`
然后运行`python tools/voc2012.py`切分为训练和验证的tfrecord数据集。
我通过阅读代码，并且debug发现，除了本人上述描述的，VOCdevkit不需要做其它数据处理，而且voc2012.py运行不报错，成功转换为tfrecord文件，其它没有改动任何代码，都是按照文档来的。
但是迁移训练时候非常慢，跑了100多轮后还是没有收敛，损失8.0左右。比着使用完整的VOC2012原始数据还要慢，请问是我数据制作有问题吗，因为我数据只是做了test.txt，train.txt，trainval.txt，val.txt切分，相比VOC2012没有做其他的处理？谢谢！

**Comment by VeeranjaneyuluToka** [comment]
I do want to check whether the way i am passing bbox labels are correct. My labels are originally in xml format xmin, ymin, xmax, ymax wrt original image size, then i am converting xml file to csv and then converting to tfrecords by following the below link https://medium.com/@omcar17/how-to-convert-xml-files-into-tfrecords-in-tensorflow2-0-86120b553f0b. As i understand, the labels needs to be converted to center_x, center_y, and w, h before we pass to YOLO architecture, but i do not understand where this centralization gets calculated in the code. Any comments on the approach that i am following ?

**Comment by hei-my** [comment]
> I do want to check whether the way i am passing bbox labels are correct. My labels are originally in xml format xmin, ymin, xmax, ymax wrt original image size, then i am converting xml file to csv and then converting to tfrecords by following the below link https://medium.com/@omcar17/how-to-convert-xml-files-into-tfrecords-in-tensorflow2-0-86120b553f0b. As i understand, the labels needs to be converted to center_x, center_y, and w, h before we pass to YOLO architecture, but i do not understand where this centralization gets calculated in the code. Any comments on the approach that i am following ?

My XML format is `<annotation>
	<folder>slider</folder>
	<filename>0.jpg</filename>
	<path>D:\data\python\slider\0.jpg</path>
	<source>
		<database>Unknown</database>
	</source>
	<size>
		<width>320</width>
		<height>160</height>
		<depth>3</depth>
	</size>
	<segmented>0</segmented>
	<object>
		<name>slider</name>
		<pose>Unspecified</pose>
		<truncated>0</truncated>
		<difficult>0</difficult>
		<bndbox>
			<xmin>169</xmin>
			<ymin>61</ymin>
			<xmax>223</xmax>
			<ymax>116</ymax>
		</bndbox>
	</object>
	<object>
		<name>slider</name>
		<pose>Unspecified</pose>
		<truncated>0</truncated>
		<difficult>0</difficult>
		<bndbox>
			<xmin>235</xmin>
			<ymin>25</ymin>
			<xmax>289</xmax>
			<ymax>80</ymax>
		</bndbox>
	</object>
</annotation>
`
`---------------------------------------------------------------------------------------------------------------------`
Did I pass the BBox tag in the right way，Here are the results of my training after several interruptions。
`Epoch 00127: saving model to /content/drive/notebook/slider/slider/checkpoint/slider_train.tf
Epoch 128/500000
14/14 [==============================] - 50s 3s/step - loss: 510.0937 - yolo_output_0_loss: 0.1397 - yolo_output_1_loss: 16.9426 - yolo_output_2_loss: 484.3006 - val_loss: 572004302848.0000 - val_yolo_output_0_loss: 0.0000e+00 - val_yolo_output_1_loss: 572004302848.0000 - val_yolo_output_2_loss: 0.0000e+00

Epoch 00128: saving model to /content/drive/notebook/slider/slider/checkpoint/slider_train.tf
Epoch 129/500000
14/14 [==============================] - 53s 4s/step - loss: 20.2879 - yolo_output_0_loss: 0.0126 - yolo_output_1_loss: 6.1573 - yolo_output_2_loss: 0.9669 - val_loss: 134044728.0000 - val_yolo_output_0_loss: 7.3212e-30 - val_yolo_output_1_loss: 134044712.0000 - val_yolo_output_2_loss: 0.0000e+00

Epoch 00129: saving model to /content/drive/notebook/slider/slider/checkpoint/slider_train.tf
Epoch 130/500000
14/14 [==============================] - 50s 3s/step - loss: 17.6409 - yolo_output_0_loss: 0.0049 - yolo_output_1_loss: 2.6769 - yolo_output_2_loss: 0.2408 - val_loss: 1120306.8750 - val_yolo_output_0_loss: 1.8937e-08 - val_yolo_output_1_loss: 1120292.0000 - val_yolo_output_2_loss: 0.0000e+00

Epoch 00130: saving model to /content/drive/notebook/slider/slider/checkpoint/slider_train.tf
Epoch 131/500000
14/14 [==============================] - 51s 3s/step - loss: 16.5327 - yolo_output_0_loss: 0.0036 - yolo_output_1_loss: 1.4005 - yolo_output_2_loss: 0.1721 - val_loss: 23186.2012 - val_yolo_output_0_loss: 3.8120e-04 - val_yolo_output_1_loss: 23171.3066 - val_yolo_output_2_loss: 0.0000e+00

Epoch 00131: saving model to /content/drive/notebook/slider/slider/checkpoint/slider_train.tf
Epoch 132/500000
14/14 [==============================] - 52s 3s/step - loss: 16.0620 - yolo_output_0_loss: 0.0032 - yolo_output_1_loss: 1.0760 - yolo_output_2_loss: 0.1279 - val_loss: 2299.1086 - val_yolo_output_0_loss: 0.0026 - val_yolo_output_1_loss: 2284.3809 - val_yolo_output_2_loss: 5.4786e-04

Epoch 00132: saving model to /content/drive/notebook/slider/slider/checkpoint/slider_train.tf
Epoch 133/500000
14/14 [==============================] - 51s 3s/step - loss: 15.6690 - yolo_output_0_loss: 0.0025 - yolo_output_1_loss: 0.8811 - yolo_output_2_loss: 0.1057 - val_loss: 314.5774 - val_yolo_output_0_loss: 0.0087 - val_yolo_output_1_loss: 299.7287 - val_yolo_output_2_loss: 0.3005
Epoch 00133: saving model to /content/drive/notebook/slider/slider/checkpoint/slider_train.tf
Epoch 134/500000
14/14 [==============================] - 50s 3s/step - loss: 15.2416 - yolo_output_0_loss: 0.0023 - yolo_output_1_loss: 0.6571 - yolo_output_2_loss: 0.0878 - val_loss: 105.2177 - val_yolo_output_0_loss: 0.0096 - val_yolo_output_1_loss: 88.6406 - val_yolo_output_2_loss: 2.2104`

**Comment by hei-my** [comment]
> I do want to check whether the way i am passing bbox labels are correct. My labels are originally in xml format xmin, ymin, xmax, ymax wrt original image size, then i am converting xml file to csv and then converting to tfrecords by following the below link https://medium.com/@omcar17/how-to-convert-xml-files-into-tfrecords-in-tensorflow2-0-86120b553f0b. As i understand, the labels needs to be converted to center_x, center_y, and w, h before we pass to YOLO architecture, but i do not understand where this centralization gets calculated in the code. Any comments on the approach that i am following ?

**Comment by VeeranjaneyuluToka** [comment]
Looks like correct as loss is keep decreasing, i have couple of questions. 1. you might be converting these xml files to tfrecords files, right? and then passing those tfrecords files to train.py? wondering where bbox coordinates scaled wrt to network input size which is 416? 2. what is your data size?

**Comment by hei-my** [comment]
> Looks like correct as loss is keep decreasing, i have couple of questions. 1. you might be converting these xml files to tfrecords files, right? and then passing those tfrecords files to train.py? wondering where bbox coordinates scaled wrt to network input size which is 416? 2. what is your data size? 
I have 1000 pictures and annotated files. Can't I just run python tools/voc2012.py directly?Is there anything else I need to change in the code?
Does the BBox coordinate scale refer to the below?Or do I need to scale it extra?If I need to scale, is there any reference for your project? 
If yes, could you please give me your contact information?
`xmin.append(float(obj['bndbox']['xmin']) / width)
            ymin.append(float(obj['bndbox']['ymin']) / height)
            xmax.append(float(obj['bndbox']['xmax']) / width)
            ymax.append(float(obj['bndbox']['ymax']) / height)`

**Comment by hei-my** [comment]
> Looks like correct as loss is keep decreasing, i have couple of questions. 1. you might be converting these xml files to tfrecords files, right? and then passing those tfrecords files to train.py? wondering where bbox coordinates scaled wrt to network input size which is 416? 2. what is your data size?

wondering where bbox coordinates scaled wrt to network input size which is 416? 
Do you mean this place? [size, size, channels]? Then in train.py, size defaults to flags.DEFINE_integer('size', 416, 'image size')?The size of my picture is 320*160 and 600*300, I need to modify the default value of size, right?
`def YoloV3Tiny(size=None, channels=3, anchors=yolo_tiny_anchors,
               masks=yolo_tiny_anchor_masks, classes=80, training=False):
    x = inputs = Input([size, size, channels], name='input')`

**Comment by hei-my** [comment]
> Looks like correct as loss is keep decreasing, i have couple of questions. 1. you might be converting these xml files to tfrecords files, right? and then passing those tfrecords files to train.py? wondering where bbox coordinates scaled wrt to network input size which is 416? 2. what is your data size?

wondering where bbox coordinates scaled wrt to network input size which is 416?
Do you mean this place? [size, size, channels]? Then in train.py, size defaults to flags.DEFINE_integer('size', 416, 'image size')?The size of my picture is 320*160 and 600*300, I need to modify the default value of size, right?
def YoloV3Tiny(size=None, channels=3, anchors=yolo_tiny_anchors, masks=yolo_tiny_anchor_masks, classes=80, training=False): x = inputs = Input([size, size, channels], name='input')

**Comment by hei-my** [comment]
> Looks like correct as loss is keep decreasing, i have couple of questions. 1. you might be converting these xml files to tfrecords files, right? and then passing those tfrecords files to train.py? wondering where bbox coordinates scaled wrt to network input size which is 416? 2. what is your data size?

wondering where bbox coordinates scaled wrt to network input size which is 416?
Do you mean this place? `[size, size, channels]`? Then in train.py, size defaults to `flags.DEFINE_integer('size', 416, 'image size')`?The size of my picture is `320*160` and `600*300`, I need to modify the default value of size, and then I don't have to change anything else, and then I can do migration training, right?
`def YoloV3(size=None, channels=3, anchors=yolo_anchors,
           masks=yolo_anchor_masks, classes=80, training=False):
    x = inputs = Input([size, size, channels], name='input')`
